src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,337 @@
+      }
+    }
+    for(auto& stencil : stencilInstantiation->getStencils()) {
+      for(const int& ID : stencilIDsToVisit) {
+        if(ID == stencil->getStencilID()) {
+          fullExtent.merge(
+              analyzeStencilExtents(stencil, stencilInstantiation->getAccessIDFromName(fieldname)));
+          if(StencilBCsApplied_.count(fieldname) == 0) {
+            StencilBCsApplied_.emplace(fieldname, std::vector<int>{stencil->getStencilID()});
+          } else {
+            StencilBCsApplied_.find(fieldname)->second.push_back(stencil->getStencilID());
+          }
+          break;
+        }
+      }
+    }
+    return fullExtent;
+  };
+
+  // Loop through all the StmtAccessPair in the stencil forward
+  for(const auto& stencilPtr : stencilInstantiation->getStencils()) {
+    Stencil& stencil = *stencilPtr;
+    DAWN_LOG(INFO) << "analyzing stencil " << stencilInstantiation->getName();
+    std::unordered_map<int, Extents> stencilDirtyFields;
+    stencilDirtyFields.clear();
+
+    for(const auto& multiStagePtr : stencil.getMultiStages()) {
+      MultiStage& multiStage = *multiStagePtr;
+      for(auto stageIt = multiStage.getStages().begin(); stageIt != multiStage.getStages().end();
+          ++stageIt) {
+        Stage& stage = (**stageIt);
+        for(const auto& domethod : stage.getDoMethods()) {
+          for(const auto& stmtAccess : domethod->getStatementAccessesPairs()) {
+            Accesses& acesses = *(stmtAccess->getAccesses());
+            const auto& allReadAccesses = acesses.getReadAccesses();
+            const auto& allWriteAccesses = acesses.getWriteAccesses();
+
+            // ReadAccesses can trigger Halo-Updates and Boundary conditions if the following
+            // criteria are fullfilled:
+            // It is a Field (ID!=-1) and we had a write before from another stencil (is in
+            // dirtyFields)
+            for(const auto& readacccess : allReadAccesses) {
+              int originalID = getOriginalID(readacccess.first);
+              if(originalID == -1)
+                continue;
+              auto idWithExtents = dirtyFields.find(originalID);
+              if(idWithExtents != dirtyFields.end()) {
+                // If the access is horizontally pointwise or it is a horizontally cached field,
+                // we do not need to trigger a BC
+                if(!readacccess.second.isHorizontalPointwise() &&
+                   !stencilInstantiation->getCachedVariableSet().count(readacccess.first)) {
+                  auto finder = allBCs.find(originalID);
+                  // Check if a boundary condition for this variable was defined
+                  if(finder != allBCs.end()) {
+                    // Create the Statement to insert the Boundary conditon into the StencilDescAST
+                    std::shared_ptr<BoundaryConditionDeclStmt> boundaryConditionCall =
+                        std::make_shared<BoundaryConditionDeclStmt>(finder->second.functor);
```

Comment:
Here we create the BoundaryConditionDeclStmt from a BoundaryCondition. Just wondering, why not creating it directly in the StencilInstantiation instead of storing BoundaryConditions
https://github.com/MeteoSwiss-APN/dawn/pull/49/files#diff-c86dcdb35edb57df3601bd8e76cc64ccR889
?